activerecord/lib/active_record/reflection.rb

```diff
@@ -23,6 +24,7 @@ def add_reflection(ar, name, reflection)
         ar.clear_reflections_cache
         name = -name.to_s
         ar._reflections = ar._reflections.except(name).merge!(name => reflection)
+        ar._plural_reflections = ar._plural_reflections.except(name).merge!(reflection.plural_name => reflection)
```

Comment:
To follow the existing pattern I think this should be `except(reflection.plural_name)`, but while we're here: I don't think the use of `except` here makes any sense? (In either this case or the line above it's mirroring.)

The first theory that comes to mind is that it dates to a time when `_reflections` was a HWIA, and `merge!` of a {string,symbol} couldn't be trusted to overwrite an existing entry of the corresponding {symbol,string}? Either way, I'm curious whether anything turns up if we just delete the call?

Reply:
Ah, yeah, I think the `except` can be skipped, but it seems that if I do, I have to use `merge` merge!` fails a number of tests seemingly finding the wrong column.

Reply:
`except` was introduced to the original line in https://github.com/rails/rails/commit/652258e41a882acccdb9a3ce211dbf356e738b28, and the test added there still fails if it's removed: the order of the keys in the hash matters.

That doesn't apply to this new hash, so just using `merge` seems fine.

Reply:
We can't use `plural_name` here, because it's affected by `pluralize_table_names`:

https://github.com/rails/rails/blob/436207aa2011463e49feb6310fa144bb0a96bf0c/activerecord/lib/active_record/reflection.rb#L354-L355

The bigger problem though is that `pluralize` and `singularize` aren't always inverses:

```ruby
"tomatoes".singularize.pluralize # => "tomatos"
```

Previously, `Salad.where(tomatoes: [...])` would find a `tomato` association, but now `_plural_reflections` will contain `tomatos` instead.

Reply:
`plural_name` is interesting... that kinda sounds so egregiously wrong as to warrant a rename.

As for imperfect inverses... I'd hope that most terms used in an application would end up with well-defined inflections just because we already have different places that inflect in different directions. But it's a good point that that's not 100%. I'm tempted to add a deprecation warning on a fallback path when both these hashes and the column name misses, and ignore the corner case of an inconsistently-inversed association that conflicts with an also-present column. 🤔

Reply:
I was able to get a plural name inconsistency, but not the one I expected,

This test passes:

<details>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :tomatos, force: true do |t|
  end
end

class Post < ActiveRecord::Base
  has_one :tomato
end

class Tomato < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_plural_name
    assert_equal "tomatoes", Post.reflections["tomato"].plural_name
  end
end
```

</details>

Maybe technically this should use `tomatos` since that is my table's name? instead it still has a plural name of `tomatoes`.

So, it seems it is possible to get an inconstancy this way. It seems `has_many` uses whatever plural version you used to define the `has_many`. So, `has_many :tomatoes` has `tomatoes` as the plural name and `has_many :tomatos` has `tomatos` as the plural name, which seems good (maybe?). I am not totally sure how bad this inconsistency is since we aren't explicitly round tripping name.

I think in most cases this should have the plural version you expect.

Reply:
Ugh, sorry @hparker - tomato/tomatoes does round-trip because it's [special-cased](https://github.com/rails/rails/blob/436207aa2011463e49feb6310fa144bb0a96bf0c/activesupport/lib/active_support/inflections.rb#L20). I tested with potato/potatoes, and then for some reason decided to use a different example without checking it (???).

Here's the case I'm thinking of - it works on main and fails on this branch.

<details>

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :salads, force: true do |t|
    t.bigint :potato_id
  end

  create_table :potatos, force: true do |t|
  end
end

class Salad < ActiveRecord::Base
  belongs_to :potato
end

class Potato < ActiveRecord::Base
  has_one :salad
end

class BugTest < Minitest::Test
  def test_association_plural_name
    potato = Potato.create!
    Salad.where(potatoes: potato).load
  end
end
```

</details>

Reply:
> we already have different places that inflect in different directions

In fact, I think there's technically an argument to be made that this is (ultimately) an improvement: my recollection claims that we prefer pluralization, and singularization is comparatively rare and better avoided (at least at the AR level -- controllers to models mapping is probably a different matter)?

Reply:
> I'm tempted to add a deprecation warning on a fallback path when both these hashes and the column name misses

I like that idea. Something like: "In Rails 7.2, singular associations will only be queryable by their plural form when an inflection is defined. To continue querying the potato association as potatoes, define an inflection that pluralizes potato to potatoes."

> and ignore the corner case of an inconsistently-inversed association that conflicts with an also-present column

That case would be when you have a `potatoes` column, a `potato` association, and a query that uses `potatoes` and currently hits the association.

It certainly seems like an odd thing to do... but maybe not beyond the realm of possibility. It's also not ambiguous - we can detect it for the low, low cost of a call to `singularize`:

```patch
diff --git a/activerecord/lib/active_record/relation/predicate_builder.rb b/activerecord/lib/active_record/relation/predicate_builder.rb
index 1a1091556e..e5f55aaed9 100644
--- a/activerecord/lib/active_record/relation/predicate_builder.rb
+++ b/activerecord/lib/active_record/relation/predicate_builder.rb
@@ -122,6 +122,10 @@ def expand_from_hash(attributes, &block)
 
               grouping_queries(queries)
             end
+          elsif table.associated_with_singular?(key.singularize)
+            ActiveSupport::Deprecation.warn(<<~MSG)
+              [...]
+            MSG
+            # treat key as an association
           else
             self[key, value]
           end
```

We'd be giving back the performance gains until after the deprecation cycle, but in exchange for covering all cases with the deprecation.

Reply:
> That case would be when you have a potatoes column, a potato association, and a query that uses potatoes and currently hits the association.

Ooh, your spelling it out here helps: we could theoretically detect this situation without runtime `singularize` by singularizing column names (at ~boot) instead.